ethclient/ethclient.go

```diff
@@ -108,10 +108,10 @@ func (ec *Client) PeerCount(ctx context.Context) (uint64, error) {
 	return uint64(result), err
 }
 
-// BlockReceipts returns the receipts of a given block number or hash
+// BlockReceipts returns the receipts of a given block number or hash.
 func (ec *Client) BlockReceipts(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) ([]*types.Receipt, error) {
 	var r []*types.Receipt
-	err := ec.c.CallContext(ctx, &r, "eth_getBlockReceipts", blockNrOrHash)
+	err := ec.c.CallContext(ctx, &r, "eth_getBlockReceipts", blockNrOrHash.String())
```

Comment:
This is highly suspect.  `BlockNrOrHash.String()` returns numbers in decimal which won't work with `eth_getBlockReceipts` as it expects an object parameter like `{"blockNumber":"0x7fffffffffffffff"}` or a hex encoded number starting with `0x` (or a block hash).

I strongly suspect this is a bug in upstream geth.  Will need to do some testing to confirm. https://github.com/ethereum/go-ethereum/pull/28087 is the PR it was changed in.  The tests for this type only cover JSON round trip, not trying to parse the output of `String()` as JSON.

Reply:
For our purposes we could just remove the `String()` call again.  It's non-standard but the object format (`{"blockNumber":"0x7fffffffffffffff"}`) works fine or geth and is what it would currently send. The execution specs say it should just be a single value that's any of:
* block number (hex encoded with `0x` prefix)
* block hash
* block label (`latest`, `pending`, `safe`, `finalized`)


Alternatively I think the right fix will be to change `BlockNumberOrHash.String()` to be:
```go
func (bnh *BlockNumberOrHash) String() string {
	if bnh.BlockNumber != nil {
		return bnh.BlockNumber.String()
	}
	if bnh.BlockHash != nil {
		return bnh.BlockHash.String()
	}
	return "nil"
}
```

The key change there is to use the `BlockNumber.String()` function and not `strconv.Itoa`.

Reply:
I've raised https://github.com/ethereum/go-ethereum/pull/28358 to fix this upstream.

Reply:
Thinking further on this, I don't think we should remove the `.String()` because we won't get conflicts on that and will almost certainly forget to add it back once the String method is fixed.  So I'm going to add my upstream fix onto this PR. We'll get merge conflicts when we merge it from upstream but should be really easy to resolve.